File: 0403-frog-jump/0403-frog-jump.py
class Solution:
    def canCross(self, stones: List[int]) -> bool:
        dp = [[-1]*2001 for _ in range(2001)]
        mp = {}
        for i in range(0, len(stones)):
            mp[stones[i]] = i
        return self.f(dp, mp, stones, len(stones), 0, 0)
    def f(self, dp, mp, stones, n, i, prevJump):
        if i == n-1: return True
        if dp[i][prevJump] != -1: return dp[i][prevJump]
        ans = False
        for k in range(prevJump-1, prevJump+2):
            if k > 0 and stones[i]+k in mp:
                ans = ans or self.f(dp, mp, stones, n, mp[stones[i]+k], k)
        dp[i][prevJump] = ans
        return ans

File: 0403-frog-jump/0403-frog-jump.java
class Solution {
    HashMap<Integer, Integer> mp = new HashMap<>();
    int[][] dp = new int[2001][2001];
    public boolean canCross(int[] stones) {
        for(int i=0; i<stones.length; i++) mp.put(stones[i], i);
        for(int[] row : dp) Arrays.fill(row, -1);
        return f(stones, stones.length, 0, 0);
    }
    boolean f(int[] stones, int n, int i, int prevJump){
        if(i == n-1) return true;
        if(dp[i][prevJump] != -1) return dp[i][prevJump] == 1;
        boolean ans = false;
        for(int k=prevJump-1; k<=prevJump+1; k++){
            if(k > 0 && mp.containsKey(stones[i]+k)){
                ans = ans || f(stones, n, mp.get(stones[i]+k), k);
            }
        }
        dp[i][prevJump] = ans ? 1 : 0;
        return ans;
    }
}